### ISSUE TITLE
Implementing on Docker?

### ISSUE THREAD

**Issue by KoalaGeo** [AstunTechnology/iso19139.gemini23#107]
Hi, 

Have you any examples of adding this plugin to a deployment on Docker? 

Can you run the COPY arguments in the dockerfile? Or is it better to run https://github.com/geonetwork/core-geonetwork/blob/3.10.x/add-schema.sh in the dockerfile? Or another way?

Many thanks, 

Edd

**Comment by archaeogeek** [comment]
Hi,

I implement it myself in docker (3.10.x) by including it in a war file and then deploying in a manner similar to the notes on deploying with a custom war here: https://github.com/geonetwork/docker-geonetwork/tree/master/4.0.0-alpha.1. The benefit of doing it that way is that I can test the build with jetty before building the war file and deploying it. I haven't tried any other approaches though.

All the best

Jo

**Comment by KoalaGeo** [comment]
Cheers for that, I'll have a go....:-)

**Comment by archaeogeek** [comment]
Great- I'll close this then!

Jo

**Comment by KoalaGeo** [comment]
Hi Jo,

Sorry please could you checkout my [gist](https://gist.github.com/KoalaGeo/b31666235b3784086bc7f8fb8643d129)

The files are copying in to the container but I still can't see any changes to Metadata & Templates > Standards....

Wanted to avoid building the war file route. 

Cheers

**Comment by archaeogeek** [comment]
I don't know that much yet about implementing any geonetwork customisations in docker other than the approach I'm using, but when I look at your gist there seems to be a discrepancy between the data_dir you're setting and the one you're copying the files into- I would check when the system is up and running where it's actually getting the schema from.

**Comment by KoalaGeo** [comment]
Hi, 

Got there in the end! https://gist.github.com/KoalaGeo/b31666235b3784086bc7f8fb8643d129 

Cheers :-)

**Comment by KoalaGeo** [comment]
The templates are all in there however when try to use to create a record theres "Create metadata error"
 
![image](https://user-images.githubusercontent.com/33657876/94158000-c2262880-fe79-11ea-9aa0-cecf36ada543.png)

**Comment by archaeogeek** [comment]
There are a number of reasons why that occurs- you need to find the log files to figure out which problem it is!

**Comment by KoalaGeo** [comment]
I'm getting the below, no prob the creating metadata using the other templates. It also won't let me duplicate the Gemini templates to make new records: 

`Impl.java:139)
        at org.hibernate.internal.AbstractSessionImpl$NonContextualJdbcConnectionAccess.obtainConnection(AbstractSessionImpl.java:380)
        at org.hibernate.engine.jdbc.internal.LogicalConnectionImpl.obtainConnection(LogicalConnectionImpl.java:228)
        ... 40 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:607)
        at org.postgresql.core.PGStream.<init>(PGStream.java:64)
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:147)
        ... 54 more`

**Comment by archaeogeek** [comment]
I've seen this before but with two different causes. I'll come back to you with details, but it will take a while, sorry!

**Comment by archaeogeek** [comment]
@KoalaGeo OK it's to do with the jar file- the one in the repository is slightly outdated. I'll have to rebuild it. I'll let you know when it's done.

**Comment by archaeogeek** [comment]
@KoalaGeo I've committed a new jar file `target/schema-iso19139.gemini23-3.7.jar` on the 310newjar branch. Could you try deploying that one and let me know how you get on please?

**Comment by KoalaGeo** [comment]
So there's two 404's on loadng the create record page, and a 400 error when actually click create. 
 
![image](https://user-images.githubusercontent.com/33657876/94831857-1cd4fc80-0405-11eb-9385-03703d897ee4.png)

Thanks for your patience on this!

**Comment by archaeogeek** [comment]
You can ignore the 404s, but we could do with knowing the response to the 400 error. I think I know what it's going to say already but the new jar file in the new branch should have sorted it :thinking:

**Comment by KoalaGeo** [comment]
This bit? 
![image](https://user-images.githubusercontent.com/33657876/94832401-c4eac580-0405-11eb-882c-27e7de71f252.png)

**Comment by KoalaGeo** [comment]
![image](https://user-images.githubusercontent.com/33657876/94832946-7558c980-0406-11eb-8410-590bf20b82be.png)

**Comment by KoalaGeo** [comment]
Added that output in comment - https://gist.github.com/KoalaGeo/b31666235b3784086bc7f8fb8643d129 

I've downloaded & copied in again - just redeploying

**Comment by archaeogeek** [comment]
OK, it's definitely the same issue that I had, and that I think you should have the fix for. The key is that the more recent jar file is 12.1kb whereas the incorrect one is only 11.5kb. For information, I might not be able to help much tomorrow, then I'm on leave all next week, so if we can't fix this today you might want to put it to one side for a bit!

**Comment by KoalaGeo** [comment]
From the newjar branch the one I've got is 11.8kb.... 
![image](https://user-images.githubusercontent.com/33657876/94835908-3c225880-040a-11eb-914b-978bf0ae2be0.png)

**Comment by KoalaGeo** [comment]
Still getting same error on after deploying. Maybe I need to learn the build from source war route...

**Comment by archaeogeek** [comment]
That's the right one- 12,095 is what mine reports in bytes. So... when you redeploy are you rebuilding the image from scratch- is it possible that the file in your image isn't getting updated?

**Comment by KoalaGeo** [comment]
I'm doing:
`FROM geonetwork:3.10.3-postgres

ENV POSTGRES_DB_HOST=postgres
ENV POSTGRES_DB_PORT=5432
ENV POSTGRES_DB_USERNAME=postgres
ENV POSTGRES_DB_PASSWORD=postgres
ENV POSTGRES_DB_NAME=geonetwork
ENV DATA_DIR=$CATALINA_HOME/webapps/geonetwork/WEB-INF/data

# Add Gemini Plugin
RUN mkdir $DATA_DIR/config/schema_plugins/iso19139.gemini23/
COPY geonetwork_customisation/iso19139.gemini23-3.10.x/src/main/plugin/iso19139.gemini23 $DATA_DIR/config/schema_plugins/iso19139.gemini23/
COPY geonetwork_customisation/iso19139.gemini23-3.10.x/target/schema-iso19139.gemini23-3.7.jar $CATALINA_HOME/webapps/geonetwork/WEB_INF/lib/

# Restart

CMD ["catalina.sh", "run"]`

**Comment by KoalaGeo** [comment]
Will cont. in a couple of weeks, thanks again for your help.

**Comment by archaeogeek** [comment]
OK sorry we couldn't get it sorted now. My gut feeling, by the way, is that the actual docker image isn't getting updated with the new file, but I'm quite new to docker myself so I was just looking it up to see if I could prove my hunch either way...

**Comment by archaeogeek** [comment]
Did you get this working?  afaict it should be fine, so the only thing I can think of doing is replicating your build process exactly to see what happens...

**Comment by KoalaGeo** [comment]
I'm going to have a go later using https://github.com/geonetwork/docker-geonetwork/blob/master/3.10.4/Dockerfile but adding before the CMD to start.  

`RUN mkdir $DATA_DIR/config/schema_plugins/iso19139.gemini23/
COPY geonetwork_customisation/iso19139.gemini23-3.10.x/src/main/plugin/iso19139.gemini23 $DATA_DIR/config/schema_plugins/iso19139.gemini23/
COPY geonetwork_customisation/iso19139.gemini23-3.10.x/target/schema-iso19139.gemini23-3.7.jar $CATALINA_HOME/webapps/geonetwork/WEB_INF/lib/
`

**Comment by KoalaGeo** [comment]
That didn't work. :-(

**Comment by archaeogeek** [comment]
OK, all very weird, since my approach worked from clean! I will try to replicate your approach and see how I get on- I might be able to do this tomorrow but otherwise it will be next week.

**Comment by archaeogeek** [comment]
Hi @KoalaGeo I've sorted it! Your dockerfile has a minor typo in it- https://github.com/BritishGeologicalSurvey/gc-geonetwork/blob/main/Dockerfile.bgsgeonetwork#L13 should read `WEB-INF` rather than `WEB_INF`

Sorry it took me so long to spot. The way I actually found it was by opening a bash prompt inside the running container using `docker exec -t -i bgs-geonetwork /bin/bash` and then checking the contents of the `WEB-INF` directory.
